analyze/aps-oag.c

```diff
@@ -303,9 +305,298 @@ CTO_NODE* schedule_rest(AUG_GRAPH *aug_graph,
+}
+
+/**
+ * Given a parent inherited instance index it returns boolean indicating if its ready to be scheduled or not
+ * @param aug_graph Augmented dependency graph
+ * @param cond current condition
+ * @param instance_groups array of <ph,ch>
+ * @param i instance index to test
+ */
+static bool parent_inherited_ready_to_go(AUG_GRAPH *aug_graph, CONDITION cond, CHILD_PHASE* instance_groups, const int i)
+{
+  // TODO: further tests may be needed
+  return group_kind_ready_to_go(aug_graph, cond, instance_groups, i);
+}
+
+/**
+ * Given a geneirc instance index it returns boolean indicating if its ready to be scheduled or not
+ * @param aug_graph Augmented dependency graph
+ * @param cond current condition
+ * @param instance_groups array of <ph,ch>
+ * @param i instance index to test
+ */
+static bool generic_instance_ready_to_go(AUG_GRAPH *aug_graph, CONDITION cond, CHILD_PHASE* instance_groups, const int i)
+{
+  CHILD_PHASE group_key = instance_groups[i];
+
+  // Its a parent inherited attribute
+  if (group_key.ph < 0 && group_key.ch == -1)
+  {
+    return parent_inherited_ready_to_go(aug_graph, cond, instance_groups, i);
+  }
+
+  // Its a child inherited attribute
+  if (group_key.ph < 0 && group_key.ch > -1)
+  {
+    return child_inherited_ready_to_go(aug_graph, cond, instance_groups, i);
+  }
+
+  // Its a parent synthesized attribute
+  if (group_key.ph >= 0 && group_key.ch == -1)
+  {
+    return parent_synthesized_ready_to_go(aug_graph, cond, instance_groups, i);
+  }
+
+  // Its a child synthesized attribute
+  if (group_key.ph >= 0 && group_key.ch > -1)
+  {
+    return child_synthesized_ready_to_go(aug_graph, cond, instance_groups, i);
+  }
+
+  fatal_error("Not sure what group CHILD_PHASE belongs to");
+
+  return true;
+}
+
+/**
+ * Recursive scheduling function
+ * @param aug_graph Augmented dependency graph
+ * @param prev previous CTO node
+ * @param cond current CONDITION
+ * @param instance_groups array of <ph,ch> indexed by INSTANCE index
+ * @param remaining count of remaining instances to schedule
+ */
+static CTO_NODE* schedule_visits(AUG_GRAPH *aug_graph, CTO_NODE* prev, CONDITION cond, CHILD_PHASE* instance_groups, int remaining)
+{
+  int i;
+  int n = aug_graph->instances.length;
+  CTO_NODE* cto_node = NULL;
+  
+  /* If nothing more to do, we are done. */
+  if (remaining == 0) return NULL;
+
+  /* Outer condition is impossible, its a dead-end branch */
+  if (CONDITION_IS_IMPOSSIBLE(cond)) return NULL;
+
+  for (i = 0; i < n; i++)
+  {
+    INSTANCE *instance = &aug_graph->instances.array[i];
+    CHILD_PHASE instance_group = instance_groups[i];
+
+    // Already scheduled then ignore
+    if (aug_graph->schedule[i] != 0) continue;
+
+    /* check to see if makes sense
+    * (No need to schedule something that
+    * occurs only in a different condition branch.)
+    */
+    if (MERGED_CONDITION_IS_IMPOSSIBLE(cond, instance_condition(instance))) return false;
```

Comment:
Not sure if "false" is the right thing.  or maybe should schedule.  WHat is the purpose of scheduling an impossible instance?